File: atlas/examples/csharp/src/App/ReadingListCommand.cs
using Atlas.Example.Lib;
using FluentValidation;
using Spectre.Console;

namespace Atlas.Example.App;

/// <summary>
/// Validates command input, creates a reading-list item, and renders its fields.
/// </summary>
public sealed class ReadingListCommand
{
    private readonly ReadingList readingList;
    private readonly IValidator<ReadingListCommandRequest> validator;

    /// <summary>
    /// Initializes the command with the reading-list behavior it coordinates.
    /// </summary>
    /// <param name="readingList">The library service that creates items.</param>
    public ReadingListCommand(ReadingList readingList)
    {
        this.readingList = readingList;
        validator = new ReadingListCommandRequestValidator();
    }

    /// <summary>
    /// Runs the command for the supplied title.
    /// </summary>
    /// <param name="title">The non-empty title supplied at the command boundary.</param>
    public void Execute(string title)
    {
        var request = new ReadingListCommandRequest(title);
        validator.ValidateAndThrow(request);
        var item = readingList.Add(request.Title);
        AnsiConsole.Write(new Rule(item.Title));
        AnsiConsole.MarkupLine($"[grey]slug:[/] {Markup.Escape(item.Slug)}");
    }

    private sealed record ReadingListCommandRequest(string Title);

    private sealed class ReadingListCommandRequestValidator : AbstractValidator<ReadingListCommandRequest>
    {
        public ReadingListCommandRequestValidator()
        {
            RuleFor(request => request.Title).NotEmpty();
        }
    }
}

File: atlas/examples/ruby/gems/app/lib/atlas_example/reading_list_command.rb
# frozen_string_literal: true

require "active_support/core_ext/string/filters"
require "pastel"
require "atlas_example/reading_list"

module AtlasExample
  # Normalizes command input and renders the item returned by the library.
  class ReadingListCommand
    # Creates the command from its library service, terminal formatter, and output boundary.
    def initialize(reading_list = ReadingList.new, pastel = Pastel.new, output = $stdout)
      @reading_list = reading_list
      @pastel = pastel
      @output = output
    end

    # Creates and prints an item for the supplied command-line title.
    def execute(title)
      item = @reading_list.add(title.squish)
      @output.puts(@pastel.bold(item.title))
      @output.puts(@pastel.dim("slug: #{item.slug}"))
      item
    end
  end
end
